File: me_daily/2025/12/15/me/me.cpp



#include <bits/stdc++.h>

struct State {
    int value;
};
using Operation = std::function<bool(State&)>; //bool f(State& s);
Operation Add(int x) {
    return [x](State& s) {
        s.value += x;
        return true;
    };
}
Operation Sub(int x) {
    return [x](State& s) {
        s.value -= x;return true;
    };
}
bool ApplyAll(State& s, const std::vector<Operation>& ops) {
    int snapshot = s.value;
    for (const auto& op : ops) {
        if(!op(s)) {
            s.value = snapshot;
            return false;
        }
    }
    return true;
}

File: me_daily/2025/12/15/me/me.go
package me

import (
	"fmt"
	"sync"
)

type State struct {
	value int
}

type Operation func(s *State) error
type SafeState struct {
	mu    sync.Mutex
	value int
}

func Add(x int) Operation {
	return func(s *State) error {
		s.value += x
		return nil
	}
}

func Sub(x int) Operation {
	return func(s *State) error {
		if s.value < x {
			return fmt.Errorf("insufficient value")
		}
		s.value -= x
		return nil
	}
}
func ApplyAll(s *State, ops []Operation) error {
	snapshot := s.value
	for _, op := range ops {
		if err := op(s); err != nil {
			s.value = snapshot
			return err
		}
	}
	return nil
}
func ApplyAllSafe(s *SafeState, ops []Operation) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	snapshot := s.value
	for _, op := range ops {
		if err := op((*State)(s)); err != nil {
			s.value = snapshot
			return err
		}
	}
	return nil
}
